core/src/main/java/com/scalar/db/service/FactoryManager.java

```diff
@@ -0,0 +1,119 @@
+package com.scalar.db.service;
+
+import com.google.common.collect.ImmutableMap;
+import com.scalar.db.api.DistributedStorage;
+import com.scalar.db.api.DistributedStorageAdmin;
+import com.scalar.db.api.DistributedStorageFactory;
+import com.scalar.db.api.DistributedTransactionAdmin;
+import com.scalar.db.api.DistributedTransactionFactory;
+import com.scalar.db.api.DistributedTransactionManager;
+import com.scalar.db.api.TwoPhaseCommitTransactionManager;
+import com.scalar.db.config.DatabaseConfig;
+import java.util.Map;
+import java.util.ServiceLoader;
+
+/**
+ * This class loads {@link DistributedStorageFactory} and {@link DistributedTransactionFactory} with
+ * {@link ServiceLoader}.
+ */
+final class FactoryManager {
+  private static final Map<String, DistributedStorageFactory> DISTRIBUTED_STORAGE_FACTORIES;
+  private static final Map<String, DistributedTransactionFactory> DISTRIBUTED_TRANSACTION_FACTORIES;
+
+  private FactoryManager() {}
+
+  static {
+    // load DistributedStorageFactory with ServiceLoader
+    ImmutableMap.Builder<String, DistributedStorageFactory> distributedStorageFactoriesBuilder =
+        ImmutableMap.builder();
+    ServiceLoader<DistributedStorageFactory> distributedStorageFactoryServiceLoader =
+        ServiceLoader.load(DistributedStorageFactory.class);
+    for (DistributedStorageFactory distributedStorageFactory :
```

Comment:
So, eventually, we are separating adaptors (providers) and loading only specified ones in META-INF?

Reply:
Yes. I'm thinking that we will move each adapter to a subproject in this project as a first step. And in that case, we specify adaptors in META-INF in each subproject.